staking.md

```diff
@@ -0,0 +1,663 @@
+   - **Parameters**:
+     - `producer`: The pubkey of the producer in Call.caller.
+     - `amount`: The amount to stake in this staking cycle.
+   - **When Callable**: Only during the **Staking Epoch**.
+   - **Constraints**:
+     - Staking made in the current epoch `N` affect the staking cycle starting in epoch `N`, with leader election in epoch  `N + 1`.
+     - **Positive `amount`**:
+       - Must have sufficient AB to cover the increase.
+     - **Minimum Stake Requirement**:
+       - After adjustment, if the participant's stake is below `MINIMUM_STAKE`, they will not be eligible for leader selection in the upcoming cycle.
+   - **Behavior**:
+     - Updates the participant's stake commitment for the next staking cycle.
+     - Recalculats LB:
+       - **LB = amount**
+     - Recalculates AB:
+       - **AB = TB - LB**
+
+
+4. **`getStakedAmount(producer:pubkey, epoch:integer)`**
+   - **Description**: Returns the amount of tokens the participant has committed to stake in the staking cycle starting at `epoch` and ending in `epoch + 3`
+   - **Parameters**:
+     - `producer`: The pubkey of the producer.
+     - `epoch`: The epoch number. (Or possibly a block height?)
+   - **When Callable**: At any time.
+
+5. **`getAvailableBalance(producer:pubkey)`**
+   - **Description**: Returns the participant's current Available Balance (AB).
+   - **Parameters**:
+     - `producer`: The pubkey of the producer.
+   - **When Callable**: At any time.
+
+---
+
+## **Explanation of Balances Across Epochs**
+
+### **Key Concepts**
+
+- **Total Balance (TB)**: Sum of all tokens deposited via `deposit()` and not yet withdrawn.
+- **Locked Balance (LB)**: The total amount of tokens locked in the current staking cycle, i.e., epochs `N` to `N + 4`.
+- **Available Balance (AB)**: Calculated as the difference between TB and LB.
+  - **AB = TB - LB**
+
+### **Lifecycle of Tokens in the Staking Contract**
+
+1. **Staking Tokens (`deposit()`):**
+   - Increases TB and AB.
+   - No immediate effect on staking commitments until `adjustStake()` (or `stake()`) is called.
+
+2. **Adjusting Stake (`adjustStake()`):**
+   - Adjusts the participant's stake commitment for the next staking cycle.
+   - **Positive Adjustment:**
+     - Decreases AB by `amount`.
+     - Increases the stake commitment for the next staking cycle by `amount`.
+   - **Negative Adjustment:**
+     - Increases AB by `-amount`.
+     - Decreases the stake commitment for the next staking cycle by `-amount`.
+   - **AB Recalculation:**
+     - **AB = TB - LB**
+
+3. **Withdrawal (`withdraw()`):**
+   - Decreases TB and AB by `amount`.
+   - Cannot affect tokens locked in the any of the 5 concurrent staking cycles.
+
+4. **Epoch Transition:**
+   - At the end of the Payout Epoch (`N + 4`):
+     - Tokens locked in the current staking cycle become available.
+     - **LB** decreases by the amount of tokens that were locked.
+     - **AB** increases accordingly.
```

Comment:
All of this guarded by some max/min-logic w.r.t. the other "active" epochs, right?

Reply:
Yes there is a note later:

### **Example with Varying Stakes Across Epochs**

If participants can adjust their stake for each epoch within the staking cycle, the LB calculation becomes the maximum over the epochs.

Suppose Alice stake different amounts in each epoch:

- **Epoch 11**: 500 tokens
- **Epoch 12**: 300 tokens
- **Epoch 13**: 400 tokens

Then in epoch 13, we are still in the cycle started in epoch 11.

- **LB = max(SS[11], SS[12], SS[13])**
- **LB = max(500,300,400) = 500 tokens**


And the description of  the adjustStake functionality.